README
This is a fork of ttyconv by Alexious Chochoulas, where I aim to port my shell script wrapper isotty (which kicks serious ass)
to C.
more to come!



			ttyconv - Convert TTY encodings

	   http://www.bedroomlan.org/~alexios/coding_ttyconv.html

  Copyright (C) 2003 Alexios Chouchoulas

  ttyconv is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 2 of the License, or
  (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program; if not, write to the Free Software
  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA


All comments are welcome.

	- Alexios Chouchoulas <alexios@bedroomlan.org>


ChangeLog
2003-08-26  Alexios Chouchoulas  <alexios@bedroomlan.org>

	* ttyconv: initial version.



src/isotty.c
/* 
   isotty.c - Convert TTY connections from one encoding to another.

   Copyright (C) 2003 Alexios Chouchoulas

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2, or (at your option)
   any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software Foundation,
   Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.  

$Id: ttyconv.c,v 1.4 2003/09/27 12:45:35 alexios Exp $

$Log: ttyconv.c,v $
Revision 1.4  2003/09/27 12:45:35  alexios
Better signal handling to deal with unusual end-of-file conditions.

Revision 1.3  2003/09/13 12:17:33  alexios
Improved (read: fixed) support for changing terminal window sizes.

Revision 1.2  2003/08/29 13:20:36  alexios
Removed superfluous debugging information.

Revision 1.1.1.1  2003/08/29 12:33:18  alexios
Initial revision.


*/

#include <stdlib.h>
#include <stdio.h>
#include <fcntl.h>
#include <stropts.h>
#include <unistd.h>
#include <string.h>
#include <signal.h>
#include <sys/types.h>
#include <sys/select.h>
#include <sys/time.h>
#include <pty.h>
#include <iconv.h>
#include "cmdline.h"
#include "isotty.h"
#include <stdlib.h>

#define max(a,b) ((a) > (b)? (a): (b))

static char const rcsid[] = "$Id: isotty.c,v 1.4 2003/09/27 12:45:35 alexios Exp $";

char *progname;

typedef struct {
	iconv_t iconv;
	char    inb  [1024];
	char    outb [8192];
	char    *readptr;
	char    *inbuf;
	char    *outbuf;
	size_t     nin;
	size_t     nout;
} translate_t;

static translate_t r2l, l2r;

static pid_t          child_pid = 0;
static struct termios tty_old_state;
static int            tty_state_set = 0;
static struct winsize winsize;
static int            tty_winsize_set = 0;
static int            master;


void stty_raw (int master)
{
	struct termios tty_state;
	
	if (tcgetattr (fileno (stdin), &tty_state) < 0) {
		fprintf (stderr,
			 "%s: warning: unable to get TTY state for standard input. "
			 "The session may start acting strange.\n", progname);
		return;
	} else {
		if (tcsetattr (master, TCSAFLUSH, &tty_state) < 0) {
			fprintf (stderr,
				 "%s: warning: unable to set TTY state for pseudoterminal. "
				 "The session may start acting strange.\n", progname);
		}

		memcpy (&tty_old_state, &tty_state, sizeof (tty_old_state));
	}

	if (ioctl (fileno (stdin), TIOCGWINSZ, &winsize)) {
		fprintf (stderr,
			 "%s: warning: unable to get window size for standard input. "
			 "The session may start acting strange.\n", progname);
		return;
	} else 	if (ioctl (master, TIOCSWINSZ, &winsize)) {
		fprintf (stderr,
			 "%s: warning: unable to get window size for standard input. "
			 "The session may start acting strange.\n", progname);
		tty_winsize_set = 1;
	}

	tty_state.c_lflag &= ~(ICANON | IEXTEN | ISIG | ECHO);
	tty_state.c_iflag &= ~(ICRNL | INPCK | ISTRIP | IXON | BRKINT);
	tty_state.c_oflag &= ~OPOST;
	tty_state.c_cflag |= CS8;
	
	tty_state.c_cc [VMIN]  = 1;
	tty_state.c_cc [VTIME] = 0;
	
	if (tcsetattr (fileno (stdin), TCSAFLUSH, &tty_state) < 0) {
		fprintf (stderr,
			 "%s: warning: unable to set TTY state for standard input. "
			 "The session may start acting strange.\n", progname);
	} else tty_state_set = 1;
}


void stty_orig (void)
{
	if (tty_state_set)
		tcsetattr (fileno (stdin), TCSAFLUSH, &tty_old_state);
}



int spawn_command ()
{
	int master;

	child_pid = forkpty (&master, NULL, NULL, NULL);
	
	if (child_pid < 0) {
		perror ("forkpty");
		exit (1);
	}

	/* This is the child process */
	
	if (child_pid == 0) {
		execvp (cmdline [0], cmdline);
		perror ("execvp()");
		exit(1);
	}

	return master;
}

int open_pty_pair (int *amaster, int *aslave)
{
	int master, slave;
	char *name;
		
	master = getpt ();
	if (master < 0) return 0;
	
	if (grantpt (master) < 0 || unlockpt (master) < 0)
		goto close_master;

	name = ptsname (master);
	fprintf(stderr, "MASTER=%s\n",name);
	if (name == NULL)
		goto close_master;
	
	slave = open (name, O_RDWR);
	if (slave == -1)
		goto close_master;
	
	if (isastream (slave)) {
		if (ioctl (slave, I_PUSH, "ptem") < 0
		    || ioctl (slave, I_PUSH, "ldterm") < 0)
			goto close_slave;
	}
	
	*amaster = master;
	*aslave = slave;
	return 1;
	
 close_slave:
	close (slave);
	
 close_master:
	close (master);
	return 0;
}


void iconv_init ()
{
	if ((l2r.iconv = iconv_open (remote,local)) == (iconv_t) -1) {
		fprintf (stderr,
			 "%s: unsupported conversion %s -> %s.\n"
			 "\nPerhaps you misspelled either of the encodings?\n"
			 "Use iconv -l to see a list of possible encodings.\n",
			 progname, local, remote);
		exit (1);
	}

	if ((r2l.iconv = iconv_open (local,remote)) == (iconv_t) -1) {
		fprintf (stderr,
			 "%s: unsupported conversion %s -> %s.\n"
			 "\nPerhaps you misspelled either of the encodings?\n"
			 "Use iconv -l to see a list of possible encodings.\n",
			 progname, remote, local);
		exit (1);
	}
}

void iconv_done ()
{
	iconv_close (l2r.iconv);
	iconv_close (r2l.iconv);
}

int translate (translate_t * t, int fdin, int fdout)
{
	int n, e;
	
	while (1) {
		n = read (fdin, t->readptr,
			  sizeof (t->inb) - (t->readptr - t->inb));
		
		/* End of file condition */
		if (n <= 0) return 1;
		t->readptr += n;
		t->nin += n;
		
		t->nout = sizeof (t->outb);
		
	try_again:
		
		n = iconv (t->iconv, &t->inbuf, &t->nin, &t->outbuf, &t->nout);
		e = errno;
		
		
		if (n < 0) {
			
			/* iconv reports an illegal sequence. Try to recover. */
			
			if (e == EILSEQ) {
				char q = '?';
				
				write (fdout, t->outb, sizeof (t->outb) - t->nout);
				write (fdout, &q, 1);
				
				t->inbuf++;
				t->nin--;
				
				t->outbuf = t->outb;
				t->nout = sizeof (t->outb);
				
				goto try_again;
				
			} else if (e == EINVAL) {
				
				/* iconv reports an incomplete sequence.  Keep
				   what we have so far and wait for the rest. */
				
				write (fdout, t->outb, sizeof(t->outb) - t->nout);
				
				memmove (t->inb, t->inbuf, t->nin);
				t->inbuf = t->inb;
				t->readptr = t->inb + t->nin;
				
				t->outbuf = t->outb;
				t->nout = sizeof(t->outb);
				
				continue;
			}
		} else {
			write (fdout, t->outb, sizeof(t->outb) - t->nout);
			
			t->outbuf = t->outb;
			t->inbuf = t->inb;
			t->readptr = t->inb;
			
			break;
		}
	}

	return 0;
}


struct winsize ws, ws0;

void child_winch (int signum, siginfo_t* foo, void* bar)
{
	if (ioctl (fileno (stdin), TIOCGWINSZ, &ws0)) return;

	if (memcmp (&ws, &ws0, sizeof (ws))) {
		if (ioctl (master, TIOCSWINSZ, &ws0) == 0) {
			kill (child_pid, SIGWINCH);
			memcpy (&ws, &ws0, sizeof (ws));
		}
	}
}


void child_graceful_exit (int signum)
{
	stty_orig();
	iconv_done();

	fprintf (stderr, "%s: session ended (by signal).\n", progname);

	exit (0);
}



void process (int master)
{
	int fdmax;
	sigset_t sigmask, orig_sigmask;
	fd_set readset, errset;
	struct sigaction act;
	struct timespec tp;

	/* Set up signal handling */


	sigemptyset( &sigmask);
	sigaddset( &sigmask, SIGWINCH);
	sigaddset( &sigmask, SIGCHLD);
	sigaddset( &sigmask, SIGPIPE);
	sigprocmask( SIG_UNBLOCK, &sigmask, &orig_sigmask);

	act.sa_flags = SA_NODEFER;
	act.sa_mask = sigmask;

	act.sa_handler = child_winch;
	sigaction (SIGWINCH, &act, NULL);

	act.sa_handler = child_graceful_exit;
	sigaction (SIGCHLD, &act, NULL);
	sigaction (SIGPIPE, &act, NULL);


	/* Calculate maximum FD for pselect(2) */

	fdmax = max (fileno (stdin), fileno (stdout));
	fdmax = max (fdmax, fileno (stderr));
	fdmax = max (fdmax, master);

	/* Initialise input and output translation buffers */

	r2l.readptr = r2l.inb;
	r2l.inbuf = r2l.inb;
	r2l.outbuf = r2l.outb;

	l2r.readptr = l2r.inb;
	l2r.inbuf = l2r.inb;
	l2r.outbuf = l2r.outb;

	/* Go. */

	while (1) {
		int sel;

		FD_ZERO (&readset);
		FD_SET (fileno (stdin), &readset);
		FD_SET (master, &readset);

		FD_ZERO (&errset);
		FD_SET (fileno (stdin), &errset);
		FD_SET (master, &errset);

		tp.tv_sec = 0L;
		tp.tv_nsec = 500000000L;

		sel = pselect (fdmax + 1, &readset, NULL, &errset, &tp, &sigmask);

		/* We timed out. */
		if ((sel == 0) && (errno == 0)) continue;

		/* Are we done yet? */
		if ((sel < 1) && (errno != EINTR) && (errno > 0)) break;

		/* Received input from the user */
		if (FD_ISSET(fileno(stdin), &readset)) {
			if (translate (&l2r, fileno (stdin), master)) break;
		}

		/* Received output from the remote end */
		if (FD_ISSET(master, &readset)) {
			if (translate (&r2l, master, fileno (stdout))) break;
		}

		/* Something nasty happened, terminate */
		if (FD_ISSET( master, &errset) ||
		    FD_ISSET( fileno(stdin), &errset)) break;
	}
}

#define BUFSZ 256
char* detect()
{
	// hide the cursor and try to set the text to invisible since the magic below produces garbage on the screen
	const char set_invisible[] = "\033[30m\033[?25l\r";
	// magic string - not to be spoken out loud
	const char magic[] = "\xc3\xb8\xa5\xc3\x84\xd9\xa7\xd8\xb8\xe0\xe0\xa9\xa9\xb8\x88\xe5\xe5\x88\xa2\xa2";
	// set the cursor to visible - otherwise the cursor may be away to long and the user may notice
	const char set_visible[]="\033[D\033[6n\033[40D                                       \033[40D\r\033[?25h";
	// erase the magic string and restore terminal to sane defaults
	const char restore[] = "\r\[A\033[20D                    \033[20D\033[0m";
	char termreply[BUFSZ];
	int i;
	FILE* ttydev = 0;
	
	// bypass stdout and stderr
	ttydev = fopen("/dev/tty", "a");
	if ( NULL == ttydev) {
		goto exit;
	}
	printf("%s", set_invisible, magic, set_invisible);
	i = 0;
	while ( c = getchar() && i < 9) {
		*termreply++;
		switch (c) {
			case '\033':;
			case '0':;
			case '1':;
			case '2':;
			case '3':;
			case '4':;
			case '5':;
			case '6':;
			case '7':;
			case '8':;
			case '9':;
			case ';': *termreply = c;break;
			case 'R': *termreply = c; i = 9;break;
			default: break;
		}
	}

	// restore terminal
	print("%s", restore);
#if 0
case "$CURPOS" in
## UTF-8 detection
## mlterm: 7
## xterm: 14
## gnome-terminal: 13
*";7R"|*";10R"|*";11R"|*";12R"|*";13R"|*";14R") HOSTENC=UTF-8 ;;
## legacy korean encoding: CP949 / UHC / JOHAB
*";18R") HOSTENC=UHC;;
*";24R") HOSTENC=BIG5;;
## tested with screen EUC-CN to UTF-8 translation:
*";16R") HOSTENC=EUC-CN;;
## following tested with gnome-terminal
*";21R") HOSTENC=EUC-KR;;
*";23R") HOSTENC=EUC-JP;;
*";28R") HOSTENC=EUC-TW;;
## 8 bit encoding, encoding detection not possible.
## putty: 17R
## generic: 20R
*";17R"|*";20R") : ;;
*) test "x$DETECTENC" = x1 && STDERR "\
terminal responded: $CURPOS
please send this response, along with your host encoding and terminal setup to <arne@users.sourceforge.net>";;
esac
#endif // 0
exit:
	
}
#define LOCALE_NAME_LENGTH 1024
int main (int argc, char **argv)
{
	progname = argv[0];
	char local_locale[LOCALE_NAME_LENGTH];
	char remote_locale[LOCALE_NAME_LENGTH];
	// detect encoding
	detect_local( &local_locale);
	detect_remote( &remote_locale, argc, &argv);

	// set environment variables
	//getenv(3)
	//setenv(3)

	cmdline_parse (argc, argv);
	master = spawn_command ();
	stty_raw (master);
	iconv_init ();

	process (master);

	iconv_done ();
	stty_orig ();
	
	fprintf (stderr, "%s: session ended.\n", progname);
	
	return EXIT_SUCCESS;
}


src/isotty.h
/* 
   ttyconv.h - Main header file.

   Copyright (C) 2002 Alexios Chouchoulas

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2, or (at your option)
   any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software Foundation,
   Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.  

$Id: ttyconv.h,v 1.1.1.1 2003/08/29 12:33:18 alexios Exp $

$Log: ttyconv.h,v $
Revision 1.1.1.1  2003/08/29 12:33:18  alexios
Initial revision.


*/


#ifndef _ISOTTY_H
#define _ISOTTY_H


#define EXIT_FAILURE 1


#include "cmdline.h"
#include <assert.h>


extern char *progname;


#endif /* _ISOTTY_H */


src/cmdline.c
/* 
   cmdline.c - Command-line parser.

   Copyright (C) 2003 Alexios Chouchoulas

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2, or (at your option)
   any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software Foundation,
   Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.  

$Id: cmdline.c,v 1.4 2004/09/07 23:17:19 alexios Exp $

$Log: cmdline.c,v $
Revision 1.4  2004/09/07 23:17:19  alexios
Updated URL and email addresses.

Revision 1.3  2003/08/29 13:33:29  alexios
Removed credit to libjeeps author (again a leftover from initialising this
project from garmintools).

Revision 1.2  2003/08/29 13:19:25  alexios
Removed a superfluous command line argument.

Revision 1.1.1.1  2003/08/29 12:33:18  alexios
Initial revision.


*/

#include <stdlib.h>
#include <stdio.h>
#include <sys/types.h>
#include <string.h>
#include "isotty.h"
#include "config.h"


static char const rcsid[] = "$Id: cmdline.c,v 1.4 2004/09/07 23:17:19 alexios Exp $";


/* Option flags and variables.  These are initialized in parse_opt.  */

char *local=NULL;		/* --local=LOCAL-ENCODING */
char *remote=NULL;		/* --remote=REMOTE-ENCODING */
char **cmdline=NULL;		/* SHELL-COMMAND */


static error_t parse_opt (int key, char *arg, struct argp_state *state);
static void show_version (FILE *stream, struct argp_state *state);


/* The argp functions examine these global variables.  */
const char *argp_program_bug_address = "<thomas.eriksson@gmail.com>";
void (*argp_program_version_hook) (FILE *, struct argp_state *) = show_version;


static char args_doc[] = "-l LOCAL-ENCODING -r REMOTE-ENCODING [SHELL-COMMAND]";


struct argp argp =
{
	options, parse_opt, args_doc,
	"Convert TTY connections from one encoding to another.\n\nA"
	"pseudo-terminal is set up with encoding REMOTE-ENCODING. The system"
	"is assumed to be in the specified LOCAL-ENCODING. The optional"
	"SHELL-COMMAND is executed. In default of a SHELL-COMMAND, the user's"
        "shell is executed (as specified in the SHELL environment variable)."
	"Input from the local side to the SHELL-COMMAND is translated from"
        "LOCAL-ENCODING to REMOTE-ENCODING. Output from the SHELL-COMMAND to"
        "the local side is translated inversely.",
	NULL, NULL, NULL
};


struct argp_option options[] =
{
	{ "remote",         'r',     "ENCODING",             0,
	  "Specify remote encoding"},
	
	{ "local",      'l',           "ENCODING",      0,
	  "Specify local encoding", 0 },
	
	{ NULL, 0, NULL, 0, NULL, 0 }
};






/* Show the version number and copyright information.  */
static void show_version (FILE *stream, struct argp_state *state) 
{
	(void) state;
	/* Print in small parts whose localizations can hopefully be copied
	   from other programs.  */
	fputs(PACKAGE" "VERSION"\n\n", stream);
	fprintf(stream, 
		"isotty by Thomas Eriksson <thomas.eriksson@gmail.com>\n"
		"Based on ttyconv 0.2.3 by Alexios Chouchoulas <alexios@bedroomlan.org>.\n"
		"Copyright (C) %s %s\n"
		"Copyright (C) %s %s\n"
		"This program is free software; you may redistribute it under the terms of\n"
		"the GNU General Public License.  This program has absolutely no warranty.\n",
		"2010", "Thomas Eriksson", 
		"2003", "Alexios Chouchoulas");
}


/* Parse a single option.  */
static error_t parse_opt (int key, char *arg, struct argp_state *state)
{
	switch (key) {
	case 'l':			/* --local */
		local = strdup (arg);
		break;
		
	case 'r':			/* --remote */
		remote = strdup (arg);
		break;
		
         case ARGP_KEY_ARG:
		 /* End processing here. */
		 cmdline = &state->argv [state->next - 1];
		 state->next = state->argc;
		 break;

	default:
		return ARGP_ERR_UNKNOWN;
	}
	
	return 0;
}

void
cmdline_parse (int argc, char ** argv)
{
	/* Parse the arguments */
	
	argp_parse (&argp, argc, argv, ARGP_IN_ORDER, NULL, NULL);

	/* Validate the arguments */
	
	if ((local == NULL) || (remote == NULL)) {
		fprintf (stderr,
			 "%s: both local and remote encodings should be specified.\n"
			 "\nUse %s --help for usage information.\n",
			 progname,progname);
		exit (1);
	}

	/* Set up the default command, if one is needed. */

	if (cmdline == NULL) {
		if (NULL == (cmdline = malloc (2 * sizeof (char *)))) {
			perror ("malloc()");
			exit (1);
		}
		
		cmdline [0] = getenv ("SHELL");
		if (NULL == cmdline [0]) {
			cmdline [0] = strdup("/bin/sh");
		}
		cmdline [1] = NULL;

	}
}


src/ttyconv.c
/* 
   ttyconv.c - Convert TTY connections from one encoding to another.

   Copyright (C) 2003 Alexios Chouchoulas

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2, or (at your option)
   any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software Foundation,
   Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.  

$Id: ttyconv.c,v 1.4 2003/09/27 12:45:35 alexios Exp $

$Log: ttyconv.c,v $
Revision 1.4  2003/09/27 12:45:35  alexios
Better signal handling to deal with unusual end-of-file conditions.

Revision 1.3  2003/09/13 12:17:33  alexios
Improved (read: fixed) support for changing terminal window sizes.

Revision 1.2  2003/08/29 13:20:36  alexios
Removed superfluous debugging information.

Revision 1.1.1.1  2003/08/29 12:33:18  alexios
Initial revision.


*/

#include <stdlib.h>
#include <stdio.h>
#include <fcntl.h>
#include <stropts.h>
#include <unistd.h>
#include <string.h>
#include <signal.h>
#include <sys/types.h>
#include <sys/select.h>
#include <sys/time.h>
#include <pty.h>
#include <iconv.h>
#include "cmdline.h"
#include "ttyconv.h"

//#define DEBUG


#define max(a,b) ((a) > (b)? (a): (b))

static char const rcsid[] = "$Id: ttyconv.c,v 1.4 2003/09/27 12:45:35 alexios Exp $";

char *progname;

typedef struct {
	iconv_t iconv;
	char    inb  [1024];
	char    outb [8192];
	char    *readptr;
	char    *inbuf;
	char    *outbuf;
	int     nin;
	int     nout;
} translate_t;

static translate_t r2l, l2r;

static pid_t          child_pid = 0;
static struct termios tty_old_state;
static int            tty_state_set = 0;
static struct winsize winsize;
static int            tty_winsize_set = 0;
static int            master;
	


void
stty_raw (int master)
{
	struct termios tty_state;
	
	if (tcgetattr (fileno (stdin), &tty_state) < 0) {
		fprintf (stderr,
			 "%s: warning: unable to get TTY state for standard input. "
			 "The session may start acting strange.\n", progname);
		return;
	} else {
		if (tcsetattr (master, TCSAFLUSH, &tty_state) < 0) {
			fprintf (stderr,
				 "%s: warning: unable to set TTY state for pseudoterminal. "
				 "The session may start acting strange.\n", progname);
		}

		memcpy (&tty_old_state, &tty_state, sizeof (tty_old_state));
	}

	if (ioctl (fileno (stdin), TIOCGWINSZ, &winsize)) {
		fprintf (stderr,
			 "%s: warning: unable to get window size for standard input. "
			 "The session may start acting strange.\n", progname);
		return;
	} else 	if (ioctl (master, TIOCSWINSZ, &winsize)) {
		fprintf (stderr,
			 "%s: warning: unable to get window size for standard input. "
			 "The session may start acting strange.\n", progname);
		tty_winsize_set = 1;
	}

	tty_state.c_lflag &= ~(ICANON | IEXTEN | ISIG | ECHO);
	tty_state.c_iflag &= ~(ICRNL | INPCK | ISTRIP | IXON | BRKINT);
	tty_state.c_oflag &= ~OPOST;
	tty_state.c_cflag |= CS8;
	
	tty_state.c_cc [VMIN]  = 1;
	tty_state.c_cc [VTIME] = 0;
	
	if (tcsetattr (fileno (stdin), TCSAFLUSH, &tty_state) < 0) {
		fprintf (stderr,
			 "%s: warning: unable to set TTY state for standard input. "
			 "The session may start acting strange.\n", progname);
	} else tty_state_set = 1;
}


void stty_orig (void)
{
	if (tty_state_set)
		tcsetattr (fileno (stdin), TCSAFLUSH, &tty_old_state);
}



int
spawn_command ()
{
	int master;

	child_pid = forkpty (&master, NULL, NULL, NULL);
	
	if (child_pid < 0) {
		perror ("forkpty");
		exit (1);
	}

	/* This is the child process */
	
	if (child_pid == 0) {
#ifdef DEBUG
		fprintf(stderr, "*** CMDLINE={\"%s\",\"%s\"}\n", cmdline[0], cmdline[1]);
#endif
		execvp (cmdline [0], cmdline);
		perror ("execvp()");
		exit(1);
	}

#ifdef DEBUG	
	fprintf (stderr, "CHILD PROCESS PID=%d, pid=%d\n", child_pid, getpid());
#endif

	return master;
}


int
open_pty_pair (int *amaster, int *aslave)
{
	int master, slave;
	char *name;
		
	master = getpt ();
	if (master < 0) return 0;
	
	if (grantpt (master) < 0 || unlockpt (master) < 0)
		goto close_master;

	name = ptsname (master);
	fprintf(stderr, "MASTER=%s\n",name);
	if (name == NULL)
		goto close_master;
	
	slave = open (name, O_RDWR);
	if (slave == -1)
		goto close_master;
	
	if (isastream (slave)) {
		if (ioctl (slave, I_PUSH, "ptem") < 0
		    || ioctl (slave, I_PUSH, "ldterm") < 0)
			goto close_slave;
	}
	
	*amaster = master;
	*aslave = slave;
	return 1;
	
 close_slave:
	close (slave);
	
 close_master:
	close (master);
	return 0;
}


void
iconv_init ()
{
	if ((l2r.iconv = iconv_open (remote,local)) == (iconv_t) -1) {
		fprintf (stderr,
			 "%s: unsupported conversion %s -> %s.\n"
			 "\nPerhaps you misspelled either of the encodings?\n"
			 "Use iconv -l to see a list of possible encodings.\n",
			 progname, local, remote);
		exit (1);
	}

	if ((r2l.iconv = iconv_open (local,remote)) == (iconv_t) -1) {
		fprintf (stderr,
			 "%s: unsupported conversion %s -> %s.\n"
			 "\nPerhaps you misspelled either of the encodings?\n"
			 "Use iconv -l to see a list of possible encodings.\n",
			 progname, remote, local);
		exit (1);
	}
}


void
iconv_done ()
{
	iconv_close (l2r.iconv);
	iconv_close (r2l.iconv);
}


int
translate (translate_t * t, int fdin, int fdout)
{
	int n, e;
	
	while (1) {
		n = read (fdin, t->readptr,
			  sizeof (t->inb) - (t->readptr - t->inb));
		
#ifdef DEBUG
		fprintf (stderr,"READ:   read(fdin, %x, %d) = %d\n",
			 t->readptr, sizeof (t->inb) - (t->readptr - t->inb), n);
#endif
		
		/* End of file condition */
		if (n <= 0) return 1;
		t->readptr += n;
		t->nin += n;
		
#ifdef DEBUG
		for (n=0; n<sizeof (t->inb); n++) {
			fprintf(stderr,"%02x ",(unsigned char)t->inb[n]);
		}
		fprintf(stderr,"\n");
#endif
		
		t->nout = sizeof (t->outb);
		
#ifdef DEBUG
		fprintf (stderr,"BEFORE: inbuf=%p, nin=%d, outbuf=%p, nout=%d\n",
			 t->inbuf, t->nin, t->outbuf, t->nout);
#endif
				
	try_again:
#ifdef DEBUG
		fprintf(stderr,"ICONV:  iconv (r2l, %x, %d, %x, %d) = ",
			t->inbuf, t->nin, t->outbuf, t->nout);
#endif
		
		n = iconv (t->iconv, &t->inbuf, &t->nin, &t->outbuf, &t->nout);
		e = errno;
		
#ifdef DEBUG
		fprintf(stderr,"%d (errno=%d)\n", n, e);
#endif
		
		if (n < 0) {
			
			/* iconv reports an illegal sequence. Try to recover. */
			
			if (e == EILSEQ) {
				char q = '?';
#ifdef DEBUG
				fprintf(stderr,
					"illegal sequence at position %d.\n",
					t->inbuf - t->inb);
#endif
				
				write (fdout, t->outb, sizeof (t->outb) - t->nout);
				write (fdout, &q, 1);
				
				t->inbuf++;
				t->nin--;
				
				t->outbuf = t->outb;
				t->nout = sizeof (t->outb);
				
				goto try_again;
				
			} else if (e == EINVAL) {
				
				/* iconv reports an incomplete sequence.  Keep
				   what we have so far and wait for the rest. */
				
#ifdef DEBUG
				fprintf(stderr,
					"incomplete sequence at position %d.\n",
					t->inbuf - t->inb);
#endif
				
				write (fdout, t->outb, sizeof (t->outb) - t->nout);
				
				memmove (t->inb, t->inbuf, t->nin);
				t->inbuf = t->inb;
				t->readptr = t->inb + t->nin;
				
				t->outbuf = t->outb;
				t->nout = sizeof (t->outb);
				
				continue;
			}
		} else {
			write (fdout, t->outb, sizeof (t->outb) - t->nout);
			
#ifdef DEBUG
			fprintf (stderr,
				 "AFTER:  inbuf=%p, nin=%d, outbuf=%p, nout=%d\n",
				 t->inbuf, t->nin, t->outbuf, t->nout);
#endif
			
			t->outbuf = t->outb;
			t->inbuf = t->inb;
			t->readptr = t->inb;
			
			break;
		}
	}

	return 0;
}


struct winsize ws, ws0;

void child_winch (int signum)
{
	if (ioctl (fileno (stdin), TIOCGWINSZ, &ws0)) return;

	if (memcmp (&ws, &ws0, sizeof (ws))) {
		if (ioctl (master, TIOCSWINSZ, &ws0) == 0) {
			kill (child_pid, SIGWINCH);
			memcpy (&ws, &ws0, sizeof (ws));
		}
	}
}


void child_graceful_exit (int signum)
{
	stty_orig();
	iconv_done();

	fprintf (stderr, "%s: session ended (by signal).\n", progname);

	exit (0);
}



void process (int master)
{
	int fdmax;
	sigset_t sigmask, orig_sigmask;
	fd_set readset, errset;
	struct sigaction act;
	struct timespec tp;

	/* Set up signal handling */


	sigemptyset (&sigmask);
	sigaddset (&sigmask, SIGWINCH);
	sigaddset (&sigmask, SIGCHLD);
	sigaddset (&sigmask, SIGPIPE);
	sigprocmask (SIG_UNBLOCK, &sigmask, &orig_sigmask);

	act.sa_handler = child_winch;
	sigaction (SIGWINCH, &act, NULL);

	act.sa_handler = child_graceful_exit;
	sigaction (SIGCHLD, &act, NULL);
	sigaction (SIGPIPE, &act, NULL);


	/* Calculate maximum FD for pselect(2) */

	fdmax = max (fileno (stdin), fileno (stdout));
	fdmax = max (fdmax, fileno (stderr));
	fdmax = max (fdmax, master);

	/* Initialise input and output translation buffers */

	r2l.readptr = r2l.inb;
	r2l.inbuf = r2l.inb;
	r2l.outbuf = r2l.outb;

	l2r.readptr = l2r.inb;
	l2r.inbuf = l2r.inb;
	l2r.outbuf = l2r.outb;

	/* Go. */

	while (1) {
		int sel;

		FD_ZERO (&readset);
		FD_SET (fileno (stdin), &readset);
		FD_SET (master, &readset);

		FD_ZERO (&errset);
		FD_SET (fileno (stdin), &errset);
		FD_SET (master, &errset);

		tp.tv_sec = 0L;
		tp.tv_nsec = 500000000L;

		sel = pselect (fdmax + 1, &readset, NULL, &errset, &tp, &sigmask);

		/* We timed out. */
		if ((sel == 0) && (errno == 0)) continue;

		/* Are we done yet? */
		if ((sel < 1) && (errno != EINTR) && (errno > 0)) break;

		/* Received input from the user */
		if (FD_ISSET (fileno (stdin), &readset)) {
#ifdef DEBUG
			fprintf (stderr, "INPUT\n");
#endif
			if (translate (&l2r, fileno (stdin), master)) break;
		}

		/* Received output from the remote end */
		if (FD_ISSET (master, &readset)) {
#ifdef DEBUG
			fprintf (stderr, "OUTPUT\n");
#endif
			if (translate (&r2l, master, fileno (stdout))) break;
		}

		/* Something nasty happened, terminate */
		if (FD_ISSET (master, &errset) ||
		    FD_ISSET (fileno (stdin), &errset)) break;
	}
}



int
main (int argc, char **argv)
{
	progname = argv[0];
	
	cmdline_parse (argc, argv);
	master = spawn_command ();
	stty_raw (master);
	iconv_init ();

	process (master);

	iconv_done ();
	stty_orig ();
	
	fprintf (stderr, "%s: session ended.\n", progname);
	
	return EXIT_SUCCESS;
}
